File: linux/src/vpnmon/history.py
"""Журнал состояния и разбор его в отрезки для графика.

Формат и логика повторяют Windows-версию, чтобы журналы с обеих платформ
читались одинаково: строки ``ГГГГ-ММ-ДД ЧЧ:ММ:СС;состояние``.

Пишется не каждая проверка, а смена состояния плюс «сердцебиение» раз в
HEARTBEAT_MINUTES. Это держит файл компактным и позволяет отличить «связь
была всё время» от «программа не работала»: разрыв между соседними записями
больше GAP_MINUTES означает отсутствие наблюдения. Так выключение машины,
сон и аварийное завершение не превращаются в ложный зелёный участок.
"""

import datetime
import os
import threading

from . import config

UNKNOWN = "idle"
UP = "up"
DOWN = "down"

HEARTBEAT_MINUTES = 5
GAP_MINUTES = 12
RETENTION_DAYS = 30

TIME_FORMAT = "%Y-%m-%d %H:%M:%S"

HISTORY_PATH = os.path.join(config.CONFIG_DIR, "history.csv")

_lock = threading.Lock()
_last_write = None
_last_state = None


class Sample(object):
    __slots__ = ("time", "state")

    def __init__(self, time, state):
        self.time = time
        self.state = state


class Segment(object):
    __slots__ = ("start", "end", "state")

    def __init__(self, start, end, state):
        self.start = start
        self.end = end
        self.state = state

    @property
    def duration(self):
        return self.end - self.start


def record(now, state):
    """Записать состояние, если оно изменилось или истёк интервал сердцебиения."""
    global _last_write, _last_state

    with _lock:
        same = state == _last_state and _last_write is not None
        if same and (now - _last_write).total_seconds() < HEARTBEAT_MINUTES * 60:
            return

        try:
            os.makedirs(config.CONFIG_DIR, exist_ok=True)
            with open(HISTORY_PATH, "a", encoding="utf-8") as fh:
                fh.write("%s;%s\n" % (now.strftime(TIME_FORMAT), state))
            _last_write = now
            _last_state = state
        except OSError:
            # Журнал вспомогательный: сбой записи не должен ронять наблюдение.
            pass


def load(since):
    """Записи начиная с since плюс последняя перед ним.

    Без предыдущей записи неизвестно состояние на левой границе диапазона.
    """
    result = []
    before = None

    try:
        with open(HISTORY_PATH, "r", encoding="utf-8") as fh:
            lines = fh.readlines()
    except OSError:
        return result

    for raw in lines:
        line = raw.strip()
        if not line or ";" not in line:
            continue
        stamp, _, state = line.partition(";")
        try:
            when = datetime.datetime.strptime(stamp, TIME_FORMAT)
        except ValueError:
            continue

        state = state.strip().lower()
        if state not in (UP, DOWN, UNKNOWN):
            state = UNKNOWN

        sample = Sample(when, state)
        if when < since:
            before = sample
        else:
            result.append(sample)

    if before is not None:
        result.insert(0, before)
    return result


def prune():
    try:
        if not os.path.exists(HISTORY_PATH):
            return
        cutoff = datetime.datetime.now() - datetime.timedelta(days=RETENTION_DAYS)

        with open(HISTORY_PATH, "r", encoding="utf-8") as fh:
            lines = fh.readlines()

        keep = []
        for raw in lines:
            line = raw.strip()
            if not line:
                continue
            stamp = line.partition(";")[0]
            try:
                if datetime.datetime.strptime(stamp, TIME_FORMAT) < cutoff:
                    continue
            except ValueError:
                pass
            keep.append(line)

        if len(keep) != len([l for l in lines if l.strip()]):
            tmp = HISTORY_PATH + ".tmp"
            with open(tmp, "w", encoding="utf-8") as fh:
                fh.write("\n".join(keep) + ("\n" if keep else ""))
            os.replace(tmp, HISTORY_PATH)
    except OSError:
        pass


def build_segments(samples, since, until):
    """Разворачивает точечные записи в непрерывную ленту без пропусков."""
    raw = []

    def add(start, end, state):
        if end > start:
            raw.append(Segment(start, end, state))

    if not samples:
        add(since, until, UNKNOWN)
        return _merge(raw)

    cursor = since
    gap = datetime.timedelta(minutes=GAP_MINUTES)

    for i, sample in enumerate(samples):
        s_start = sample.time
        s_end = samples[i + 1].time if i + 1 < len(samples) else until
        s_end = min(s_end, until)

        # Запись подтверждает состояние лишь на ближайшие GAP_MINUTES;
        # дальше — только если пришла следующая запись.
        confirmed_until = s_start + gap
        known_end = min(s_end, confirmed_until)

        start = max(s_start, cursor)
        if start > cursor:
            add(cursor, start, UNKNOWN)

        add(start, max(start, min(known_end, until)), sample.state)

        if known_end < s_end:
            add(max(known_end, since), min(s_end, until), UNKNOWN)

        cursor = max(cursor, min(s_end, until))

    if cursor < until:
        add(cursor, until, UNKNOWN)

    return _merge(raw)


def _merge(segments):
    merged = []
    for seg in segments:
        if merged:
            last = merged[-1]
            if last.state == seg.state and last.end >= seg.start:
                if seg.end > last.end:
                    last.end = seg.end
                continue
        merged.append(seg)
    return merged


def summarize(segments):
    """Итоги по ленте: сколько было связи, сколько её не было, обрывы."""
    up = datetime.timedelta()
    down = datetime.timedelta()
    idle = datetime.timedelta()
    outages = []

    for seg in segments:
        if seg.state == UP:
            up += seg.duration
        elif seg.state == DOWN:
            down += seg.duration
            outages.append(seg)
        else:
            idle += seg.duration

    observed = up + down
    availability = (
        up.total_seconds() / observed.total_seconds() * 100.0
        if observed.total_seconds() > 0
        else None
    )
    longest = max((o.duration for o in outages), default=datetime.timedelta())

    return {
        "up": up,
        "down": down,
        "idle": idle,
        "observed": observed,
        "availability": availability,
        "outages": outages,
        "longest": longest,
    }


def format_duration(delta):
    total = int(delta.total_seconds())
    if total < 1:
        return "—"
    if total < 60:
        return "%d сек" % total
    if total < 3600:
        return "%d мин" % (total // 60)
    if total < 86400:
        return "%d ч %d мин" % (total // 3600, (total % 3600) // 60)
    return "%d д %d ч %d мин" % (
        total // 86400,
        (total % 86400) // 3600,
        (total % 3600) // 60,
    )

File: src/History.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;

namespace VpnConnectMonitoring
{
    public enum VpnState
    {
        Unknown = 0,   // наблюдение не велось: программа не работала, пауза,
                       // вне расписания или выключено
        Up = 1,
        Down = 2
    }

    public class HistorySample
    {
        public DateTime Time;
        public VpnState State;
    }

    public class Segment
    {
        public DateTime Start;
        public DateTime End;
        public VpnState State;

        public TimeSpan Duration
        {
            get { return End - Start; }
        }
    }

    // Журнал состояний в %APPDATA%\VpnConnectMonitoring\history.csv.
    //
    // Пишем не каждую проверку, а смену состояния плюс «сердцебиение» раз в
    // HeartbeatMinutes. Это держит файл компактным и одновременно позволяет
    // отличить «связь была всё время» от «программа не работала»: разрыв между
    // соседними записями больше GapMinutes означает, что наблюдения не было —
    // так корректно отображаются выключение компьютера, сон и аварийное
    // завершение, после которых закрывающей записи не остаётся.
    public static class History
    {
        public const int HeartbeatMinutes = 5;
        public const int GapMinutes = 12;
        public const int RetentionDays = 30;

        const string TimeFormat = "yyyy-MM-dd HH:mm:ss";

        static DateTime lastWrite = DateTime.MinValue;
        static VpnState lastState = VpnState.Unknown;
        static readonly object gate = new object();

        public static string FilePath
        {
            get { return Path.Combine(Config.Dir, "history.csv"); }
        }

        public static void Record(DateTime now, VpnState state)
        {
            lock (gate)
            {
                bool sameState = state == lastState && lastWrite != DateTime.MinValue;
                if (sameState && (now - lastWrite).TotalMinutes < HeartbeatMinutes)
                    return;

                try
                {
                    Directory.CreateDirectory(Config.Dir);
                    File.AppendAllText(FilePath,
                        now.ToString(TimeFormat, CultureInfo.InvariantCulture)
                            + ";" + Code(state) + Environment.NewLine,
                        Encoding.UTF8);

                    lastWrite = now;
                    lastState = state;
                }
                catch
                {
                    // Журнал — вспомогательная функция; сбой записи не должен
                    // ломать основную работу наблюдателя.
                }
            }
        }

        static string Code(VpnState s)
        {
            if (s == VpnState.Up) return "up";
            if (s == VpnState.Down) return "down";
            return "idle";
        }

        static VpnState Parse(string s)
        {
            if (s == "up") return VpnState.Up;
            if (s == "down") return VpnState.Down;
            return VpnState.Unknown;
        }

        // Возвращает записи начиная с from, плюс последнюю запись перед from —
        // без неё неизвестно состояние на левой границе диапазона.
        public static List<HistorySample> Load(DateTime from)
        {
            List<HistorySample> result = new List<HistorySample>();
            HistorySample before = null;

            try
            {
                if (!File.Exists(FilePath))
                    return result;

                foreach (string raw in File.ReadAllLines(FilePath, Encoding.UTF8))
                {
                    string line = raw.Trim();
                    if (line.Length == 0)
                        continue;

                    int sep = line.IndexOf(';');
                    if (sep <= 0)
                        continue;

                    DateTime t;
                    if (!DateTime.TryParseExact(line.Substring(0, sep), TimeFormat,
                            CultureInfo.InvariantCulture, DateTimeStyles.None, out t))
                        continue;

                    HistorySample s = new HistorySample();
                    s.Time = t;
                    s.State = Parse(line.Substring(sep + 1).Trim().ToLowerInvariant());

                    if (t < from)
                        before = s;
                    else
                        result.Add(s);
                }
            }
            catch
            {
                return result;
            }

            if (before != null)
                result.Insert(0, before);

            return result;
        }

        public static void Prune()
        {
            try
            {
                if (!File.Exists(FilePath))
                    return;

                DateTime cutoff = DateTime.Now.Date.AddDays(-RetentionDays);
                string[] lines = File.ReadAllLines(FilePath, Encoding.UTF8);

                List<string> keep = new List<string>();
                for (int i = 0; i < lines.Length; i++)
                {
                    string line = lines[i].Trim();
                    if (line.Length == 0)
                        continue;

                    int sep = line.IndexOf(';');
                    DateTime t;
                    if (sep > 0 && DateTime.TryParseExact(line.Substring(0, sep), TimeFormat,
                            CultureInfo.InvariantCulture, DateTimeStyles.None, out t)
                        && t < cutoff)
                    {
                        continue;
                    }
                    keep.Add(line);
                }

                if (keep.Count != lines.Length)
                    File.WriteAllLines(FilePath, keep.ToArray(), Encoding.UTF8);
            }
            catch { }
        }

        // Разворачивает точечные записи в непрерывную ленту отрезков,
        // покрывающую весь диапазон [from, to] без пропусков.
        public static List<Segment> BuildSegments(List<HistorySample> samples,
            DateTime from, DateTime to)
        {
            List<Segment> raw = new List<Segment>();

            if (samples == null || samples.Count == 0)
            {
                Add(raw, from, to, VpnState.Unknown);
                return Merge(raw);
            }

            DateTime cursor = from;

            for (int i = 0; i < samples.Count; i++)
            {
                DateTime sStart = samples[i].Time;
                DateTime sEnd = (i + 1 < samples.Count) ? samples[i + 1].Time : to;
                if (sEnd > to) sEnd = to;

                // Запись подтверждает состояние лишь на ближайшие GapMinutes;
                // дальше — только если пришла следующая запись.
                DateTime confirmedUntil = sStart.AddMinutes(GapMinutes);
                DateTime knownEnd = sEnd < confirmedUntil ? sEnd : confirmedUntil;

                DateTime a = Max(sStart, cursor);
                if (a > cursor)
                    Add(raw, cursor, a, VpnState.Unknown);

                Add(raw, a, Max(a, Min(knownEnd, to)), samples[i].State);

                if (knownEnd < sEnd)
                    Add(raw, Max(knownEnd, from), Min(sEnd, to), VpnState.Unknown);

                cursor = Max(cursor, Min(sEnd, to));
            }

            if (cursor < to)
                Add(raw, cursor, to, VpnState.Unknown);

            return Merge(raw);
        }

        static void Add(List<Segment> list, DateTime start, DateTime end, VpnState state)
        {
            if (end <= start)
                return;

            Segment s = new Segment();
            s.Start = start;
            s.End = end;
            s.State = state;
            list.Add(s);
        }

        static List<Segment> Merge(List<Segment> input)
        {
            List<Segment> result = new List<Segment>();
            for (int i = 0; i < input.Count; i++)
            {
                if (result.Count > 0)
                {
                    Segment last = result[result.Count - 1];
                    if (last.State == input[i].State && last.End >= input[i].Start)
                    {
                        if (input[i].End > last.End)
                            last.End = input[i].End;
                        continue;
                    }
                }
                result.Add(input[i]);
            }
            return result;
        }

        static DateTime Min(DateTime a, DateTime b) { return a < b ? a : b; }
        static DateTime Max(DateTime a, DateTime b) { return a > b ? a : b; }
    }
}
